test/test.js

```diff
@@ -1,61 +1,210 @@
 import chai from 'chai'
+import sinon from 'sinon';
 import nock from 'nock'
 import es6promise from 'es6-promise'
 const expect = chai.expect
 
 es6promise.polyfill()
-import jsonFetch from '..'
+import jsonFetch, {retriers} from '..'
 
 describe('jsonFetch',() => {
-  it('resolves with json body for 200-level status codes', () => {
-    nock('http://www.test.com')
-      .get('/products/1234')
-      .reply(200, {name: 'apple'})
-    return jsonFetch('http://www.test.com/products/1234').then((response) => {
-      expect(response.body).to.deep.equal({name: 'apple'})
-      expect(response.status).to.equal(200)
-      expect(response.statusText).to.equal('OK')
-      expect(response.headers).to.be.ok
+  describe('single request with no retry', () => {
+    it('resolves with json body for 200-level status codes', () => {
+      nock('http://www.test.com')
+        .get('/products/1234')
+        .reply(200, {name: 'apple'})
+      return jsonFetch('http://www.test.com/products/1234').then((response) => {
+        expect(response.body).to.deep.equal({name: 'apple'})
+        expect(response.status).to.equal(200)
+        expect(response.statusText).to.equal('OK')
+        expect(response.headers).to.be.ok
+      })
     })
-  })
 
-  it('resolves with undefined for 404 status code', () => {
-    nock('http://www.test.com')
-      .get('/products/1234')
-      .reply(404, 'not found')
-    return jsonFetch('http://www.test.com/products/1234').then((response) => {
-      expect(response.body).to.equal(undefined)
-      expect(response.status).to.equal(404)
-      expect(response.statusText).to.equal('Not Found')
-      expect(response.headers).to.be.ok
+    it('resolves with undefined for 404 status code', () => {
```

Comment:
Hmm, we _could_. Some http libraries work that way -- throw an error for statusCodes >= 400. I think there's some real merit and precedence behind that approach.

My personal experience is that for some endpoints, a 200 status can contain an error or a 419 can be rate limiting, or other surprising status codes can have strange semantics in a given context. Because of those experiences, I tend to think of endpoint error semantics as highly application specific. In my personally preferred approach, any http response is a success for the http client and its the caller's responsibility (aka an SDK) to determine whether the http response represents a success or failure with its greater knowledge of the context.

So, @dlau and I simply reformated this line, and we didn't change the semantics of this test or the behavior of jsonFetch. If we did treat specific response status codes as failures, I expect that would be a major version bump -- which is totally fine!

I guess, the most intuitive answer to me is: "what does fetch do?" According to [MDN](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch#Checking_that_the_fetch_was_successful):

> A fetch() promise will reject with a TypeError when a network error is encountered, although this usually means permission issues or similar — a 404 does not constitute a network error

I did a very quick look at node-fetch, and I found [this test case](https://github.com/bitinn/node-fetch/blob/master/test/test.js#L844-L849):

```js
fetch(url, opts).then(function(res) {
			expect(res.status).to.equal(404);
})
```

Anyway, I'm happy to chat in person in more depth if you'd like? Let me know!